ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -43,26 +40,85 @@ func ReadCheckpointV6(dir string, fileName string) ([]*trie.MTrie, error) {
 	return tries, nil
 }
 
-func filePathHeader(dir string, fileName string) string {
+func filePathCheckpointHeader(dir string, fileName string) string {
 	return path.Join(dir, fileName)
 }
 
 func filePathSubTries(dir string, fileName string, index int) (string, string, error) {
-	if index < 0 || index > 15 {
+	if index < 0 || index > (subtrieCount-1) {
 		return "", "", fmt.Errorf("index must be between 1 to 16, but got %v", index)
 	}
 	subTrieFileName := fmt.Sprintf("%v.%v", fileName, index)
 	return path.Join(dir, subTrieFileName), subTrieFileName, nil
 }
 
 func filePathTopTries(dir string, fileName string) (string, string) {
-	topTriesFileName := fmt.Sprintf("%v.%v", fileName, 16)
+	topTriesFileName := fmt.Sprintf("%v.%v", fileName, subtrieCount)
 	return path.Join(dir, topTriesFileName), fileName
 }
 
-// TODO: to add
-func readHeader(filePath string) ([]uint32, uint32, error) {
-	return make([]uint32, 16), 0, nil
+// readCheckpointHeader takes a file path and returns subtrieChecksums and topTrieChecksum
+// any error returned are exceptions
+func readCheckpointHeader(filepath string) ([]uint32, uint32, error) {
+	closable, err := os.Open(filepath)
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not open file %v: %w", filepath, err)
+	}
+	defer func(f *os.File) {
+		f.Close()
+	}(closable)
+
+	var bufReader io.Reader = bufio.NewReaderSize(closable, defaultBufioReadSize)
+	reader := NewCRC32Reader(bufReader)
+
+	// read the magic bytes and header
+	version, err := readVersion(reader)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if version != VersionV6 {
+		return nil, 0, fmt.Errorf("wrong version: %v", version)
+	}
+
+	// read the subtrie level
+	subtrieLevel, err := readSubtrieLevel(reader)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// read subtrie checksums
+	subtrieCount := subtrieCountByLevel(subtrieLevel)
+	subtrieChecksums := make([]uint32, subtrieCount)
+	for i := 0; i < subtrieCount; i++ {
+		sum, err := readCRC32Sum(reader)
+		if err != nil {
+			return nil, 0, fmt.Errorf("could not read %v-th subtrie checksum from checkpoint header: %w", i, err)
+		}
+		subtrieChecksums[i] = sum
+	}
+
+	// read top level trie checksum
+	topTrieChecksum, err := readCRC32Sum(reader)
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not read checkpoint top level trie checksum in chechpoint summary: %w", err)
+	}
+
+	// calculate the actual checksum
+	actualSum := reader.Crc32()
+
+	// read the stored checksum, and compare with the actual sum
+	expectedSum, err := readCRC32Sum(reader)
```

Comment:
Yeah, that makes sense. 

But I have another point:

[In this PR](https://github.com/onflow/flow-go/pull/3335/files#diff-2ff6507ee260a1e37156b4cae4cc259e7688bbd0007993b20fa193824073105aR224), as you see, I'm actually getting rid of the bufReader variable. Instead, I just pass the created bufReader directly to the NewCRC32Reader.

Why?

Because I ran into a bug shown as below, where I should be using the `CRC32Reader`, however used `bufReader`, which is a bug that caused some data not seen by the crc reader to compute the correct checksum:
```
// bad
_, err = readVersion(MagicBytesCheckpointSubtrie, bufReader)
// good
_, err = readVersion(MagicBytesCheckpointSubtrie, reader)
```

The reason I was running into this bug without realizing is that both bufReader and reader are available in the context. That's why I avoided creating the bufReader variable. 

Now back to here, yes, we wasted a little effort to compute a checksum that won't be used, but the wasted effort is so little, that worth the trade off for simplicity and safety.

Reply:
I understand your point about using one reader for simplicity.

> Now back to here, yes, we wasted a little effort to compute a checksum that won't be used, but the wasted effort is so little, that maybe worth the trade off for simplicity and safety?

My main concern was about the correctness rather than the small wasted effort.  

It would helpful to document that no additional checksum computation is needed after the final computation step.

Reply:
Agreed. Added comments